argon2/src/memory_view.rs

```diff
@@ -0,0 +1,118 @@
+//! This module encapsulates the unsafe code for accessing memory in parallel.
+
+// TODO(tarcieri): safe parallel implementation
+// See: https://github.com/RustCrypto/password-hashes/issues/154
+
+use crate::{instance::SYNC_POINTS, Block};
+
+use core::slice;
+
+#[cfg(feature = "parallel")]
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+#[derive(Debug)]
+pub struct MemoryView<'a> {
+    memory: &'a mut [Block],
+    slice: usize,
+    lane_count: usize,
+    segment_length: usize,
+
+    #[cfg(feature = "parallel")]
+    cur_lane: AtomicUsize,
+
+    #[cfg(not(feature = "parallel"))]
+    cur_lane: usize,
+}
+
+impl<'a> MemoryView<'a> {
+    pub fn new(memory: &'a mut [Block], slice: usize, lane_count: usize) -> Self {
+        assert!(slice < SYNC_POINTS);
+        assert_ne!(lane_count, 0);
+
+        let segment_count = lane_count * SYNC_POINTS;
+        assert!(memory.len() >= segment_count);
+        assert_eq!(memory.len() % segment_count, 0);
+
+        let segment_length = memory.len() / segment_count;
+
+        Self {
+            memory,
+            slice,
+            lane_count,
+            segment_length,
+
+            #[cfg(feature = "parallel")]
+            cur_lane: AtomicUsize::new(0),
+
+            #[cfg(not(feature = "parallel"))]
+            cur_lane: 0,
+        }
+    }
+
+    pub fn lane(&self, lane: usize) -> (&'a [Block], &'a [Block]) {
+        assert!(lane < self.lane_count);
+
+        let base_ptr = self.memory.as_ptr();
+
+        unsafe {
+            let before_ptr = base_ptr.add(lane * self.lane_length());
+            let before_len = self.slice * self.segment_length;
+            let before_segment = slice::from_raw_parts(before_ptr, before_len);
+
+            let offset = before_len + self.segment_length;
+            let after_ptr = before_ptr.add(offset);
+            let after_len = self.lane_length() - offset;
+            let after_segment = slice::from_raw_parts(after_ptr, after_len);
+
+            (before_segment, after_segment)
+        }
```

Comment:
Is there something I'm missing that necessitates unsafe code here? Why can't this be safely written as:

```suggestion
            let before_base = &self.memory[(lane * self.lane_length()..];
            let before_len = self.slice * self.segment_length;
            let before_segment = &before_base[..before_len];

            let offset = before_len + self.segment_length;
            let after_base = &self.memory[offset..]
            let after_len = self.lane_length() - offset;
            let after_segment = &after_base[..after_len];

            (before_segment, after_segment)
```

Reply:
You're right, however the `after_base` needs to apply the offset to `before_base` instead of `self.memory`.